### PR TITLE
update combat ai behaviour; more variety

### PR DESCRIPTION
**Pull request by mrcheko**
3 stages of combat (for AI): 
far away pathgrid following
close-up when shortcutting right to player (and in the end can start attack)
melee when using kind of tactic with dodging and different kinds of attack types (no combos:( )
Added case for distant combatants(archers, throwers)

Almost all behaviour depends on reaction (mTimerReact) updates period (pathfinding is here, too)
Also path now rebuilds only when target has moved on some threshold distance

### CODE REVIEW

File: apps/openmw/mwmechanics/actors.cpp [hunk 1]
```diff
@@ -190,7 +191,7 @@ bool disintegrateSlot (MWWorld::Ptr ptr, int slot, float disintegrate)
                     && LOS
                     )
                 {
-                    creatureStats.getAiSequence().stack(AiCombat("player"));
+                    creatureStats.getAiSequence().stack(AiCombat(MWBase::Environment::get().getWorld()->getPlayer().getPlayer()));
```
Comment (ghost) [inline apps/openmw/mwmechanics/actors.cpp:213]: There is now a new method World::getPlayerPtr, so you also do not need to include ../mwworld/player anymore ;)

**Comment by zinnschlag** [comment]
@mrcheko That does not merge cleanly.
@scrawl I wasn't aware of this function. I don't think it is a good idea. World is already a big class and adding convenience functions will only bloat it further. If people find the double getPlayer call to bothersome, maybe we can add an implicit conversion operator to Ptr to the MWWorld::Player class? I think this is one of the few places where having an implicit conversion operator will not cause any harm.

**Comment by ghost** [comment]
The reason for this is not really the double getPlayer() (although this kinda impairs readability), but to cut down the number of includes.

**Comment by zinnschlag** [comment]
Kinda the same argument, but with convenience include reduction instead of convenience function call reduction.
One could argue with compile time of course, but player.hpp itself is small and it includes little that ptr.hp does not include anyway. Well, except for refdata.hpp, which actually should not be included in player.hpp in the first place (probably some leftover from earlier versions).

**Comment by ghost** [comment]
Cutting down the number of interdependencies is not just about convenience, it's also an important design choice.
95% of all code interacting with the player only needs the MWWorld::Ptr and cares about nothing else. Typically something like "doX(Ptr = player)" and checks such as "ptr == world->getPlayerPtr()". Therefore, it makes sense to forward declare MWWorld::Player in the World header, since it contains specifics that are actually only relevant for the player (birthsign, etc) which 95% of code does not care about. getPlayerPtr() is different in that it only allows access to the player-agnostic features of the player object.

(In fact there is even more code not covered by this that just does getHandle() == "player", which should be changed btw - string literal checks are bad...)

I agree World is bloated, although that is kinda unavoidable. Any game engine is bound to have a large World class.
Perhaps we should outsource some functionality that does not belong in world. In particular, creating separate interfaces for Physics and Rendering components might be helpful. Yes, most of rendering / physics is done by World and does not have to be exposed. There are exceptions however and having to add those as World methods such as playing a video, fading the screen etc which clearly have nothing to do with the World itself, definitely adds to the bloat.

**Comment by ghost** [comment]
Preliminary list of "probably-shouldn't-be-in-world" methods:

// Render
            virtual void renderPlayer() = 0;
            virtual MWRender::Animation\* getAnimation(const MWWorld::Ptr &ptr) = 0;
            virtual void playVideo(const std::string& name, bool allowSkipping) = 0;
            virtual void stopVideo() = 0;
            virtual void frameStarted (float dt, bool paused) = 0;
            virtual OEngine::Render::Fader\* getFader() = 0;
            virtual void getTriangleBatchCount(unsigned int &triangles, unsigned int &batches) = 0;
            virtual bool toggleRenderMode (RenderMode mode) = 0;

// Physics
            virtual MWWorld::Ptr  getFacedObject() = 0;
            virtual std::pairMWWorld::Ptr,Ogre::Vector3 getHitContact(const MWWorld::Ptr &ptr, float distance) = 0;
            virtual void queueMovement(const MWWorld::Ptr &ptr, const Ogre::Vector3 &velocity) = 0;
            virtual bool toggleCollisionMode() = 0;
            virtual void processChangedSettings (const Settings::CategorySettingVector& settings) = 0;
            virtual bool getPlayerStandingOn (const MWWorld::Ptr& object) = 0; ///< @return true if the player is standing on \a object
            virtual bool getActorStandingOn (const MWWorld::Ptr& object) = 0; ///< @return true if any actor is standing on \a object
            virtual bool getLOS(const MWWorld::Ptr& npc,const MWWorld::Ptr& targetNpc) = 0;
            virtual void enableActorCollision(const MWWorld::Ptr& actor, bool enable) = 0;
            virtual bool isFlying(const MWWorld::Ptr &ptr) const;
            virtual bool isSlowFalling(const MWWorld::Ptr &ptr) const;
            ///Is the head of the creature underwater?
            virtual bool isSubmerged(const MWWorld::Ptr &object) const;
            virtual bool isSwimming(const MWWorld::Ptr &object) const;
            virtual bool isUnderwater(const MWWorld::Ptr::CellStore\* cell, const Ogre::Vector3 &pos) const;
            virtual bool isOnGround(const MWWorld::Ptr &ptr) const;

Camera control (render?):
            virtual void togglePOV() = 0;
            virtual void togglePreviewMode(bool enable) = 0;
            virtual bool toggleVanityMode(bool enable) = 0;
            virtual void allowVanityMode(bool allow) = 0;
            virtual void togglePlayerLooking(bool enable) = 0;
            virtual void changeVanityModeScale(float factor) = 0;
            virtual bool vanityRotateCamera(float \* rot) = 0;
            virtual void setCameraDistance(float dist, bool adjust = false, bool override = true)=0;

**Comment by zinnschlag** [comment]
You have a point regarding most code using World only requires the Ptr and not the Player class. I wasn't aware that this is the case.

As for World bloat, I am somewhat more concerned with the World implementation doing too much, because the primary purpose of this class is to manage and to delegate. The interface bloat is indeed unavoidable to some degree,

I don't think I agree with your entire list, in particular regarding separate interfaces for physics and rendering. The physics are part of the world and the idea behind having rendering managed by World is that any change to the World will automatically be represented in the rendering.
Maybe we should consider a limited interface for rendering at least that deals with aspects of rendering that are not related to the actual world.
Anyway, these are tasks for post 1.0 development. I think the state of World is still acceptable for the time being as long as we pay attention to not making it worse than it currently is.

**Comment by ghost** [comment]
I get several warnings like this
In file included from /home/scrawl/Dev/openmw/apps/openmw/mwmechanics/aisequence.cpp:11:0:
/home/scrawl/Dev/openmw/apps/openmw/mwmechanics/aicombat.hpp:49:17: warning: ‘void MWMechanics::determineAttackType(const MWWorld::Ptr&, MWMechanics::Movement&)’ declared ‘static’ but never defined [-Wunused-function]
     static void determineAttackType(const MWWorld::Ptr& actor, MWMechanics::Movement &movement);
                 ^
/home/scrawl/Dev/openmw/apps/openmw/mwmechanics/aicombat.hpp:51:17: warning: ‘void MWMechanics::chooseBestAttack(const ESM::Weapon_, MWMechanics::Movement&)’ declared ‘static’ but never defined [-Wunused-function]
     static void chooseBestAttack(const ESM::Weapon_ weapon, MWMechanics::Movement &movement);
                 ^

**Comment by ghost** [comment]
Still getting the exact same warnings.

**Comment by ghost** [comment]
I see you added another commit. Still getting the same warning though.
The declaration for determineAttackType and chooseBestAttack should be completely removed from aicombat.hpp.

**Comment by ghost** [comment]
And a style suggestion, for internal non-member functions you might want to prefer an anonymous namespace.

File: apps/openmw/mwmechanics/aicombat.cpp [hunk 2]
```diff
@@ -26,124 +30,314 @@
...
+        {
+            //special run attack; it shouldn't affect melee combat tactics
+            if(actor.getClass().getMovementSettings(actor).mPosition[1] == 1)
             {
-                mTimer = mTimer + duration;
+                //check if actor can overcome the distance = distToTarget - attackerWeapRange
+                //less than in time of playing weapon anim from 'start' to 'hit' tags (t_swing)
+                //then start attacking
+                float speed1 = cls.getSpeed(actor);
+                float speed2 = cls.getSpeed(mTarget);
```
Comment (ghost) [inline apps/openmw/mwmechanics/aicombat.cpp]: This should be mTarget.getClass().getSpeed(mTarget), or you will get bad casts if actor class is not the same as target class.

**Comment by ghost** [comment]
Almost there. There's another class mismatch in aicombat.cpp:264 and 265 (overlooked that one)

**Comment by ghost** [comment]
This seems to break creature combat, which was just merged into master.
Creatures just try walking through me and don't attack at all.

**Comment by ghost** [comment]
Something else I noticed, StartCombat seems to be broken.
Try "startcombat, player" with one of the ordinators in toddtest. It'll start hitting thin air and the ordinator next to him instead of charging the player.
In master this is working. Hmm..

**Comment by mrcheko** [comment]
it seems like in OpStartCombat actor starts the battle against himself :)
I'll look now what's going on with creatures.

**Comment by ghost** [comment]
Ok, great. Try using Steam Centurion for testing (some have too large bounding box and can't get close enough to you, see http://bugs.openmw.org/issues/1126 )

**Comment by mrcheko** [comment]
there was problem with random negative values of weapon ranges for creatures. 
But I can't test startcombat because I get 'no implicit refernce' error though believe it works now

**Comment by ghost** [comment]
To set the implicit reference you need to click on the NPC while the console is opened.

**Comment by mrcheko** [comment]
startcombat is now OK.

**Comment by ghost** [comment]
You've got my thumbs up for the merge now.

**Comment by mrcheko** [comment]
I understood that it was a bit early to post that bug but here it is:
http://postimg.org/image/ceurs5aw3/
2/3 centurion attacks have that stretching (just check in vanilla)
It is on the Nif level as I understand.

**Comment by ghost** [comment]
Yes, it's a GeomMorpherController.
Works properly in OpenMW for me: http://scrawl.bplaced.net/temp/screenshot2014-01-18_15_07_50.png
Maybe try if morphing works at all in your build. Do you have pulsating netch air sacks?

**Comment by ghost** [comment]
Oh nevermind! I think this was the fix (not in zini's master yet): https://github.com/scrawl/openmw/commit/5049f3b1346ee80eb9f0bede5b24eb5329d7e976
I totally forgot that I had merged that in while I was testing your changes.
See if cherry picking that commit resolves the animation problem for you.

**Comment by mrcheko** [comment]
yes, it indeed works. hehe, but why?

**Comment by ghost** [comment]
It looks like only keyframe controllers support groups at the moment.

**Comment by mrcheko** [comment]
There was a bug when ai was backtracking on path rebuildings, when target moving somewhere, due to applying of the closest point as the path start (that was mentioned by gus in last comment here http://bugs.openmw.org/issues/1102). You actually can see that backtracking even now but this occure only when the current path is empty and ai builds it again. To resolve this one (which left) we should think of choosing the start point of the path more sophisticatedly.

Removed dodging for creatures (they are not that smart).

Suggestion for using aifollow package instead of just close-up shortcutting and for the areas with no pathpoints.

**Comment by ghost** [comment]
AiFollow is actually something completely different, because it will make the actor join the target actor's party (i.e. fight with him against others), so this is not an option here.

**Comment by mrcheko** [comment]
Yes, but I thought that we can reuse that target tracking system which is implemented in
https://github.com/gugus/openmw/commit/2d66b2c4fa7ca03a096a3be54fc6e19d6cdc8f83. 
And logic of party joinings won't be in AiFollow class itself, right?

**Comment by ghost** [comment]
I don't think it is a good idea to cross-reference AI packages like that. The only place where this makes sense is to spawn an AICombat package in AIFollow if the target is attacked.
If you want to reuse that code, then put it in a public function, in PathFinder maybe.

**Comment by mrcheko** [comment]
This is essentially the same logic: 

> to spawn an AICombat package in AIFollow if the target is attacked

and to spawn AIFollow in AICombat if attacker should follow target, just different point of view.
And moving that function in PathFinder is really makes sense.

**Comment by mrcheko** [comment]
I believe this will improve architecture a bit: now all mechanics of setting attack type (for NPCs, creatures, magic spells, even player's attack type handling) are located in 1 function CharacterController::determineAttackType.
Also there is one more pathfinding optimisation in aicombat.cpp, although I think we should think of something more global and sophisticated in  this area (parallel with obstacles avoiding).

**Comment by ghost** [comment]
I agree about moving chop/slash/thrust detection out of input manager.

The spell casting change makes no sense though. Actually the spell casting stance is not necessarily connected to attacking, so having it set as an "attack type" is wrong.
The spell range is only needed by the character controller for playing the correct animation and should not be exposed through CreatureStats. Actually, a spell doesn't even have a "range" as such - the range is per effect, and each effect can have a different range.

**Comment by ghost** [comment]
Also I think creatures should play attack1-3 randomly and not connected to slash/thrust/chop movement directions.

**Comment by ghost** [comment]
Perhaps add a determineAttackType method to MWWorld::Class, which would choose randomly for creatures, and work with the movement settings for NPCs.
That way you could also get rid of the isNpc checks, and use polymorphism instead.

**Comment by mrcheko** [comment]
> The spell range is only needed by the character controller for playing the correct animation and should not be exposed through CreatureStats.

It also needed in AIcombat: it matters if spell is on-touch or on-target. And also on-self spells will have a little bit different behaviour of previous two types (just imagine using of restoration magic during combat).

> Actually, a spell doesn't even have a "range" as such - the range is per effect, and each effect can have a different range.

You know it better, I only need to know how to determine if spell is on- touch/target/self for aicombat. 

> Also I think creatures should play attack1-3 randomly and not connected to slash/thrust/chop movement directions.

It's just a matter of choice: the one who will have a good idea can easily alter it since it is located in one function.
At the moment it actually behaves 90% equally as if it was done without connection to movement (there are small moves triggered in aicombat, that's 10%:) ).
And yes, maybe it would be good to move it to MWWorld::Class but we have different determining for creatures/npcs for physical combat and equal for magic combat.

**Comment by ghost** [comment]
> It also needed in AIcombat: it matters if spell is on-touch or on-target. And also on-self spells will have a little bit different behaviour of previous two types (just imagine using of restoration magic during combat).

Spell casting AI is an entirely different matter. You're not going to just select a random spell, check its range and then cast it. You need to go through all its effects anyway, to decide if it is a useful spell for this situation or not. And this is done before deciding if you even want to cast a spell in the first place, so it can't use determineAttackType.

> And yes, maybe it would be good to move it to MWWorld::Class but we have different determining for creatures/npcs for physical combat and equal for magic combat.

As I said, the magic portion does not belong in the function at all.

**Comment by ghost** [comment]
Come to think of it, CreatureStats::mAttackType can be removed entirely, now that it is no longer set from InputManager.

**Comment by mrcheko** [comment]
> Perhaps add a determineAttackType method to MWWorld::Class

I didn't find a way to organically add it there.

> CreatureStats::mAttackType can be removed entirely

It's still needed in hit(ptr, int attackType) and we can't access CharacterController::mAttackType from ::npc/::creature level.

**Comment by ghost** [comment]
> It's still needed in hit(ptr, int attackType)

And that is exactly what the "int attackType" parameter is for.

**Comment by ghost** [comment]
To clarify: Animation::handleTextKey calls :npc/creature::hit and determines the attack type by checking which animation is playing.

**Comment by mrcheko** [comment]
Yes, I know that. But how are you going to eliminate CreatureStats::mAttackType

**Comment by ghost** [comment]
Because all references to getAttackType have already been removed at this point.
That means calls to setAttackType are redundant and mAttackType should be removed too.

File: apps/openmw/mwmechanics/character.cpp [hunk 3]
```diff
@@ -1319,4 +1307,34 @@ void CharacterController::updateVisibility()
...
 }
 
+void CharacterController::determineAttackType()
+{
+    float * move = mPtr.getClass().getMovementSettings(mPtr).mPosition;
+    
+    if (move[0] && !move[1]) //sideway
+    {
+        mPtr.getClass().getCreatureStats(mPtr).setAttackType(MWMechanics::CreatureStats::AT_Slash);
+        if(mPtr.getClass().isNpc())
```
Comment (ghost) [inline apps/openmw/mwmechanics/character.cpp]: Some creatures can now also carry weapons. This needs to be merged with master and then the isNpc lines changed to hasInventoryStore.

**Comment by mrcheko** [comment]
check it now

**Comment by ghost** [comment]
Doesn't compile.

/home/scrawl/Dev/openmw/apps/openmw/mwclass/creature.cpp: In member function ‘virtual void MWClass::Creature::hit(const MWWorld::Ptr&, int) const’:
/home/scrawl/Dev/openmw/apps/openmw/mwclass/creature.cpp:285:24: error: ‘AT_Chop’ is not a member of ‘MWMechanics::CreatureStats’
/home/scrawl/Dev/openmw/apps/openmw/mwclass/creature.cpp:287:29: error: ‘AT_Slash’ is not a member of ‘MWMechanics::CreatureStats’
/home/scrawl/Dev/openmw/apps/openmw/mwclass/creature.cpp:289:29: error: ‘AT_Thrust’ is not a member of ‘MWMechanics::CreatureStats’
/home/scrawl/Dev/openmw/apps/openmw/mwclass/npc.cpp: In member function ‘virtual void MWClass::Npc::hit(const MWWorld::Ptr&, int) const’:
/home/scrawl/Dev/openmw/apps/openmw/mwclass/npc.cpp:521:24: error: ‘AT_Chop’ is not a member of ‘MWMechanics::CreatureStats’
/home/scrawl/Dev/openmw/apps/openmw/mwclass/npc.cpp:523:29: error: ‘AT_Slash’ is not a member of ‘MWMechanics::CreatureStats’
/home/scrawl/Dev/openmw/apps/openmw/mwclass/npc.cpp:525:29: error: ‘AT_Thrust’ is not a member of ‘MWMechanics::CreatureStats’

**Comment by mrcheko** [comment]
fixed

**Comment by ghost** [comment]
Ok, I think it is good to merge now.

Minor issue (can be fixed later): creatures shouldn't move around while attacking.
My suggestion would be to disable the movement code for creatures, and handle the playing of attack1-3 animations randomly (not based on movement). After all, those are unique animations and don't have anything to do with chop/slash/thrust anyway.

**Comment by zinnschlag** [comment]
That doesn't merge cleanly.

**Comment by mrcheko** [comment]
I missed what gugus achieved by this

> mMovement.mRotation[2] = 10*(Ogre::Degree(zAngle).valueRadians()-pos.rot[2]);

smoother rotations?
But anyway the old aicombat was inconsistent and unflexible.

**Comment by ghost** [comment]
1. smoother rotations, 2. movement settings do not rot rotate if the actor is paralyzed or incapacitated, etc.

So this change should be adopted.

**Pull request merged**